activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -39,8 +39,17 @@ def load_records_in_batch(loaders)
 
           def load_records_for_keys(keys, &block)
             if association_key_name.is_a?(Array)
-              or_scope = keys.map { |values_set| scope.klass.where(association_key_name.zip(values_set).to_h) }.inject(&:or)
-              scope.merge(or_scope)
+              zipped_keys = keys.flat_map { |values_set|
+                association_key_name.zip(values_set).to_h
+              }
+
+              query_constraints = zipped_keys.reduce({}) { |hsh, zipped|
```

Comment:
It looks like we should be able to avoid iterating over `keys` for the second time here and build the resulting hash during the iteration above if we used an an approach like:

```ruby
  query_constraints = Hash.new { |hsh, key| hsh[key] = Set.new }

  keys.each_with_object(query_constraints) do |values_set, constraints|
    association_key_name.zip(values_set).each do |key_name, value|
      constraints[key_name] << value
    end
  end
```

I benchmarked it and the "push to a set" solution seems to be quicker most likely due to not performing the second iteration but I'm mainly interested in improving readability, if possible